### ISSUE TITLE
Remove unused test case in twisted.test.test_reflect

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6856]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)| @multani reported|
|-|-|
|Trac ID|trac#6856|
|Type|enhancement|
|Created|2013-11-29 11:44:49Z|
|Branch|https://github.com/twisted/twisted/tree/remove-dead-code-from-test_reflect-6856|

```twisted.test.test_reflect.ImportHooksLookupTests``` is a test case which isn't used at the moment in Twisted code case and has a ```skip``` marker since r35441. It depends on the ```ihooks``` module which has been removed in Python 3.

This is something which should have been removed as part of [#6689](https://github.com/twisted/twisted/issues/6689) but we somehow missed it.
It actually raises a ```DeprecationWarning``` in [#5929](https://github.com/twisted/twisted/issues/5929) if tests are executed with Python 2.x and the ```-3``` flag.

I think it's pretty safe to remove this class and move forward on [#5929](https://github.com/twisted/twisted/issues/5929) then.

Attachments:

* [6856-remove-unused-testcase-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/33b/33bf9f10d6b9af2fd5679b1b3e4463c23be41b9d/dc3446eb7de177a22615b1c1e340cf0bc47be6d2.patch) (1767 bytes) - added by multani on 2013-11-29 11:48:24Z - 
* [6856-remove-unused-testcase-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/33b/33bf9f10d6b9af2fd5679b1b3e4463c23be41b9d/1e2e5221bf6155e6993e10aff0d09bf580036615.patch) (1594 bytes) - added by multani on 2013-11-29 11:51:49Z - Empty file and git :(
* [6856-remove-unused-testcase-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/33b/33bf9f10d6b9af2fd5679b1b3e4463c23be41b9d/1f3b050eda42fdca7f6d305690283de4f0a72e9a.patch) (1703 bytes) - added by multani on 2013-11-29 17:29:30Z - Actually, I can track empty files with Git o/

<details><summary>Searchable metadata</summary>

```
trac-id__6856 6856
type__enhancement enhancement
reporter__multani multani
priority__normal normal
milestone__ 
branch__branches_remove_dead_code_from_test_reflect_6856 branches/remove-dead-code-from-test_reflect-6856
branch_author__hynek hynek
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1385725489000000 1385725489000000
changetime__1390825141000000 1390825141000000
version__None None
owner__hynek hynek

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to @multani|
|-|-|

Hmm.  I'm not sure I follow the logic in the ticket description.  Perhaps you can expand on it somewhat?

I tried removing the skip from `ImportHooksLookupTests` and noticed that the tests all pass on Python 2.7.  It seems to only be broken on Python 2.6.  Perhaps some of the earlier porting work was overly zealous in introducing more skips - or perhaps someone has since fixed the ihooks bug in Python 2.7 (but skipped fixing it in 2.6?).

Regardless, considering that `ihooks` still exists on Python 2.6 and Python 2.7, what is the justification for deleting this test case?  ihooks being removed in Python 3 doesn't seem to be sufficient.  If I'm overlooking something, sorry - please correct me!

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_3"></a>|@multani commented|
|-|-|

[spinning my brain up to remember what was this thing about...]

Replying to [exarkun](#note_2):
> Hmm.  I'm not sure I follow the logic in the ticket description.  Perhaps you can expand on it somewhat?
> 
> I tried removing the skip from `ImportHooksLookupTests` and noticed that the tests all pass on Python 2.7.  It seems to only be broken on Python 2.6.  Perhaps some of the earlier porting work was overly zealous in introducing more skips - or perhaps someone has since fixed the ihooks bug in Python 2.7 (but skipped fixing it in 2.6?).
> 
> Regardless, considering that `ihooks` still exists on Python 2.6 and Python 2.7, what is the justification for deleting this test case?  ihooks being removed in Python 3 doesn't seem to be sufficient.  If I'm overlooking something, sorry - please correct me!

So, I clearly underlooked this test case, since it actually has tests attached to it by subclassing `LookupsTestCase` from `twisted.python.test.test_reflectpy3`, which for some reasons I missed.

I also tried to remove the `skip()` attributes from the class, and tests pass both on Python 2.6 and Python 2.7...

So.
What should I do with this?

* I can let the skip attribute, and everything "is" fine on all Python versions, although we don't know if it actually works. We'll still have the warning that ihooks has been deprecated in Python 3.0 if the tests are run from Python 2.7 with `-3` (which was the reason for creating this ticket in the first place, see [#5929](https://github.com/twisted/twisted/issues/5929)#comment:32 ) and this is basically just dead code (what's the point of keeping it if it's always skipped?)

* I can remove the skip attribute on Python 2.x, since it seems to work, and add it only on Python 3.x if `ihooks` isn't available. We'll still have the aforementioned warning (which is supposed to be on the review checklist for Python 3 patches), but at least this test will have some meaning where there are executed.

If I go the first way, I'll just close this ticket and put a comment back into [[#5929](https://github.com/twisted/twisted/issues/5929)](https://github.com/twisted/twisted/issues/5929) which refers to here. Otherwise, if I go with the second way, I guess I'll make one more patch on #5929 to add this compatibility check (unless I'm being suggested otherwise).

Thoughts?

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_5"></a>|@hynek set owner to @hynek|
|-|-|

So let’s wrap this up a bit:

 1. `ihooks` are completely **undocumented**, the only docs on it I could found is the [list of undocumented modules](http://docs.python.org/release/2.7.6/library/undoc.html)  (I didn’t know of their existence until I read through this ticket).  I don’t see any value in testing explicitly compatibility with an obscure and ostensibly buggy stdlib module.
 1. It apparently is **flaky** enough to fail on 2.6 for one of you, but to pass for the other. Because of 1., I have no reason to assume that it won’t exercise any flakiness on 2.7 too. Flaky tests are the worst; the headache from Win32 buildbots is bad enough.
 1. This test is **dead code** ATM, functioning at best as a reminder to not use `ihooks`.  I don’t think that’s the job of a test case. I would suggest to add documentation but ihooks is just too obscure to litter our docs with explicit docs about it.

Long story short, I don’t think this blob of dead code adds enough value to be useful to be had in Twisted; we have too much baggage code to care for already.

If itamar says `ihooks` is broken since 2.6 I would go and kill it off.

As for the patch, there is an empty line missing between the imports and the first class; no need to fix that, I’m mentioning it just to remind you of the Twisted Coding Standard[tm]. :)

You don’t need to do anything else unless more comments are added here or I got something wrong.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_6"></a>|@hynek commented|
|-|-|

(In [41425]) Branching to remove-dead-code-from-test_reflect-6856.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_7"></a>|@hynek set status to `closed`|
|-|-|

(In [41429]) Merge remove-dead-code-from-test_reflect-6856

Author: multani
Reviewer: hynek
Fixes: [#6856](https://github.com/twisted/twisted/issues/6856)

twisted.test.test_reflect.ImportHooksLookupTests was skipped unconditionally
because the rather obscure stdlib module ihooks was broken since 2.6.  It's
either dead code or a flaky test, so we removed it.
